tiatoolbox/models/engine/engine_abc.py

```diff
@@ -1,21 +1,586 @@
 """Defines Abstract Base Class for TIAToolbox Model Engines."""
+from __future__ import annotations
+
+import json
 from abc import ABC, abstractmethod
+from pathlib import Path
+from typing import TYPE_CHECKING, NoReturn
+
+import numpy as np
+import pandas as pd
+import torch
+import tqdm
+from torch import nn
+
+from tiatoolbox import logger
+from tiatoolbox.models.architecture import get_pretrained_model
+from tiatoolbox.models.dataset.dataset_abc import PatchDataset
+from tiatoolbox.models.models_abc import load_torch_model, model_to
+
+if TYPE_CHECKING:  # pragma: no cover
+    import os
+
+    from torch.utils.data import DataLoader
+
+    from tiatoolbox.annotation import AnnotationStore
+
+    from .io_config import ModelIOConfigABC
+
+
+def prepare_engines_save_dir(
+    save_dir: os | Path | None,
+    len_images: int,
+    *,
+    patch_mode: bool,
+    overwrite: bool,
+) -> Path | None:
+    """Create directory if not defined and number of images is more than 1.
+
+    Args:
+        save_dir (str or Path):
+            Path to output directory.
+        len_images (int):
+            List of inputs to process.
+        patch_mode(bool):
+            Whether to treat input image as a patch or WSI.
+        overwrite (bool):
+                Whether to overwrite the results. Default = False.
+
+    Returns:
+        :class:`Path`:
+            Path to output directory.
+
+    """
+    if patch_mode is True:
+        if save_dir is not None:
+            save_dir.mkdir(parents=True, exist_ok=overwrite)
+        return save_dir
+
+    if save_dir is None:
+        if len_images > 1:
+            msg = (
+                "More than 1 WSIs detected but there is no save directory provided."
+                "Please provide a 'save_dir'."
+            )
+            raise OSError(msg)
+        return (
+            Path.cwd()
+        )  # save the output to current working directory and return save_dir
+
+    if len_images > 1:
+        logger.info(
+            "When providing multiple whole-slide images / tiles, "
```

Comment:
```suggestion
            "When providing multiple whole-slide images / visual fields, "
```

Reply:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/635/commits/d970f49baeb08badedd2d1d214354348b1ffd6cd